src/composer.ts

```diff
@@ -0,0 +1,497 @@
+import {
+  TransactionSigner,
+  TransactionWithSigner,
+  isTransactionWithSigner,
+} from './signer';
+import {
+  OnApplicationComplete,
+  SuggestedParams,
+} from './types/transactions/base';
+
+export type ABIArgument = ABIValue | TransactionWithSigner;
+
+/** Represents the output from a successful ABI method call. */
+export interface ABIResult {
+  /** The TxID of the transaction that invoked the ABI method call. */
+  txID: string;
+  /**
+   * The raw bytes of the return value from the ABI method call. This will be empty if the method
+   * does not return a value (return type "void").
+   */
+  rawReturnValue: Uint8Array;
+  /**
+   * The return value from the ABI method call. This will be undefined if the method does not return
+   * a value (return type "void"), or if the SDK was unable to decode the returned value.
+   */
+  returnValue?: ABIValue;
+  /** If the SDK was unable to decode a return value, the error will be here. */
+  decodeError?: Error;
+}
+
+export enum AtomicTransactionComposerStatus {
+  /** The atomic group is still under construction. */
+  BUILDING,
+
+  /** The atomic group has been finalized, but not yet signed. */
+  BUILT,
+
+  /** The atomic group has been finalized and signed, but not yet submitted to the network. */
+  SIGNED,
+
+  /** The atomic group has been finalized, signed, and submitted to the network. */
+  SUBMITTED,
+
+  /** The atomic group has been finalized, signed, submitted, and successfully committed to a block. */
+  COMMITTED,
+}
+
+/** A class used to construct and execute atomic transaction groups */
+export class AtomicTransactionComposer {
+  /** The maximum size of an atomic transaction group. */
+  static MAX_GROUP_SIZE: number = 16;
+
+  private status = AtomicTransactionComposerStatus.BUILDING;
+  private transactions: TransactionWithSigner[] = [];
+  private methodCalls: Map<number, ABIMethod> = new Map();
+  private signedTxns: Uint8Array[] = [];
+  private txIDs: string[] = [];
+
+  /**
+   * Get the status of this composer's transaction group.
+   */
+  getStatus(): AtomicTransactionComposerStatus {
+    return this.status;
+  }
+
+  /**
+   * Get the number of transactions currently in this atomic group.
+   */
+  count(): number {
+    return this.transactions.length;
+  }
+
+  /**
+   * Create a new composer with the same underlying transactions. The new composer's status will be
+   * BUILDING, so additional transactions may be added to it.
+   */
+  clone(): AtomicTransactionComposer {
+    const theClone = new AtomicTransactionComposer();
+
+    theClone.transactions = this.transactions.map(({ txn, signer }) => ({
+      // not quite a deep copy, but good enough for our purposes (modifying txn.group in buildGroup)
+      txn: Transaction.from_obj_for_encoding({
+        ...txn.get_obj_for_encoding(),
+        // erase the group ID
+        grp: undefined,
+      }),
+      signer,
+    }));
+    theClone.methodCalls = new Map(this.methodCalls);
```

Comment:
It seems to me that you deep-copied the transactions, erased group-id and signed transactions (if they exists), and deep-copied the methodCalls (I am not sure on that one)?

Reply:
That's correct, the Map constructor performs a deep copy (well, deep enough for our purposes)